### ISSUE TITLE
Tall style: Initializers should always start in the same line as the variable declaration

### ISSUE THREAD

**Issue by rrousselGit** [dart-lang/dart_style#1526]
Hello!  
I've been playing around with the new formatting, and encountered one formatting change I think is a downgrade:

```dart
  final character = await repository.fetchCharacter(
    id,
    cancelToken: cancelToken,
  );
```

Now becomes:

```dart
  final character =
      await repository.fetchCharacter(id, cancelToken: cancelToken);
```

What I really dislike here is how the variable initializer is in the next line as the variable declaration.
The whole statement doesn't fit in a single line. So IMO we should use trailing commas here, instead of a newline.

This is consistent with how a different scenario I encountered is formatting.
Before, we had:

```dart
final character =
    FutureProvider.autoDispose.family<Character, String>((ref, id) async {
});
```

But with the new formatting, we now have:

```dart
final character = FutureProvider.autoDispose.family<Character, String>((
  ref,
  id,
) async {
});
```

IMO this is an improvement. The initializer is in the same line as the variable declaration. But it's inconsistent with the previous snippet to me.

**Comment by rrousselGit** [comment]
Another example I faced:

```dart
    final charactersResponse = await repository.fetchCharacters(
      offset: meta.page * kCharactersPageLimit,
      limit: kCharactersPageLimit,
      nameStartsWith: meta.name,
      cancelToken: cancelToken,
    );
```

now becomes:

```dart
      final charactersResponse =
          await repository.fetchCharacters(
            offset: meta.page * kCharactersPageLimit,
            limit: kCharactersPageLimit,
            nameStartsWith: meta.name,
            cancelToken: cancelToken,
          );
```

Here, I see no purpose in that newline. It just adds one extra level of indentation.

**Comment by rrousselGit** [comment]
Redirecting factory constructors seem to be facing a similar issue:

```dart
@freezed
class CharacterOffset with _$CharacterOffset {
  factory CharacterOffset({
    required int offset,
    @Default('') String name,
  }) = _CharacterOffset;
}
```

Becomes:

```dart
@freezed
class CharacterOffset with _$CharacterOffset {
  factory CharacterOffset({required int offset, @Default('') String name}) =
      _CharacterOffset;
}
```

Overall it seems to be a theme where a "statement" tries to fit the left operand of the `=` operator in a single line, even if it means placing the right one on a new line.

**Comment by rrousselGit** [comment]
Another example that I think is a downgrade:

```diff
-  final content = json.decode(
-    await rootBundle.loadString('assets/configurations.json'),
-   ) as Map<String, Object?>;
+  final content =
+      json.decode(await rootBundle.loadString('assets/configurations.json'))
+          as Map<String, Object?>;
```

**Comment by rrousselGit** [comment]
Is this caused by `await`?   
I found a similar scenario, but the formatting is different.

This:

```dart
    final results =
        annotationsOf(element, throwOnUnresolved: throwOnUnresolved);
```

Became:

```dart
    final results = annotationsOf(
      element,
      throwOnUnresolved: throwOnUnresolved,
    );
```

That makes sense to me. The "old" syntax is my mistake. I forgot to add a trailing comma, and would've added it if I spotted the issue.

But that's at odds with how some earlier examples are formatted.

**Comment by munificent** [comment]
Yes, the first two examples of yours and the last one are because of `await`. I filed #1531 for that.

Redirecting factory constructors are an interesting one. I think people generally prefer to split between a declaration and its body versus splitting the parameter list. Like:

```dart
// Better:
function(int argument, int another) =>
    someLongBodyExpression;

// Worse:
function(
  int argument,
  int another,
) => someLongBodyExpression;
```

And redirecting factory constructors pretty much follow that too. We could tweak it, but I suspect it's honestly not even worth it given how rare they are.

The other remaining example is because of the `as` expression. We do block formatting of assignment initializers if the RHS is one of a few recognized "block formattable" types. That's mainly collection literals and function calls with non-empty argument lists. Currently, we don't treat any binary operators as block formattable even if their left operand is. We *could* do that, but I suspect it would some cases better but more cases worse.

For what it's worth, this doesn't look unpalatable to me:

```dart
  final content =
     json.decode(await rootBundle.loadString('assets/configurations.json'))
         as Map<String, Object?>;
```

It may be a downgrade when you're looking at the diff versus the old formatter. But if you had written that code from scratch and then run the new formatter on it, I'm not sure if you would have even noticed what it did at all.

But I can look at tweaking the formatter and see how it looks across a wide corpus if we allow block formatting `as` expressions when the LHS is block formattable and maybe some other binary operators too (`is`?). I suspect it will be a net loss, but it's worth trying it out on a corpus and seeing.

**Comment by rrousselGit** [comment]
> Like:
> 
> ```dart
> // Better:
> function(int argument, int another) =>
>     someLongBodyExpression;
> 
> // Worse:
> function(
>   int argument,
>   int another,
> ) => someLongBodyExpression;
> ```
> 
> And redirecting factory constructors pretty much follow that too. We could tweak it, but I suspect it's honestly not even worth it given how rare they are.

I _never_ use => when the result is in a different line.
`=> expression;` on the same line is the only variant I'd accept in a codebase.

If anything, I'd much rather write:

```dart
function(int arg, ...) {
  return expression;
}
```

To me, refactoring => to {return} is very much like adding a trailing comma, but for closures.
In fact, I'd kind of like if the formatter automatically refactored to {} here.

**Comment by rrousselGit** [comment]
> For what it's worth, this doesn't look unpalatable to me:
> 
> ```dart
>   final content =
>      json.decode(await rootBundle.loadString('assets/configurations.json'))
>          as Map<String, Object?>;
> ```
> 
> It may be a downgrade when you're looking at the diff versus the old formatter. But if you had written that code from scratch and then run the new formatter on it, I'm not sure if you would have even noticed what it did at all.

This is 100% something I'd spot.

My eyes really dislike when there's a newline after the `=`. Any formatting that does that would immediately be refactored such that it no longer happens.

For instance, I could write:

```dart
final content = json.decode(...);
content as Map<...>
```

I prefer the repetition of the `content` identifier over splitting the declaration and initializer.
That's how much I dislike the split.

**Comment by munificent** [comment]
Hi, @rrousselGit. With #1687 done, I went through the examples. Here is what the formatter produces now:

```dart
// 1:
  final character = await repository.fetchCharacter(
    id,
    cancelToken: cancelToken,
  );

// 2:
      final charactersResponse = await repository.fetchCharacters(
        offset: meta.page * kCharactersPageLimit,
        limit: kCharactersPageLimit,
        nameStartsWith: meta.name,
        cancelToken: cancelToken,
      );

// 3:
@freezed
class CharacterOffset with _$CharacterOffset {
  factory CharacterOffset({required int offset, @Default('') String name}) =
      _CharacterOffset;
}

// 4:
    final response = await _get(
      'characters',
      queryParameters: <String, Object?>{
        'offset': offset,
        if (limit != null) 'limit': limit,
        if (cleanNameFilter != null && cleanNameFilter.isNotEmpty)
          'nameStartsWith': cleanNameFilter,
      },
      cancelToken: cancelToken,
    );

// 5:
  final content =
     json.decode(await rootBundle.loadString('assets/configurations.json'))
         as Map<String, Object?>;

// 6:
    final results = annotationsOf(
      element,
      throwOnUnresolved: throwOnUnresolved,
    );

// 7:
        final newListeners = List<_Listener<T>?>.filled(
          _listeners.length * 2,
          null,
        );
```

Of those, 1, 2, 4, 6, and 7 are what you prefer and what the formatter now produces.

For 3, the redirecting constructor, I think most users would rather split at the `=` than the parameter list, so I'm happy with the output here.

For 5, it splits that way because the outermost expression is `as`, not the JSON call. Binary operators have a general rule (which is true in almost every formatter) that if an operand contains a newline, then the operator is split too. That's what's going on here, and the output seems reasonable to me.

Thank you for all the feedback and concrete examples. Real-world code like this is very helpful, especially as the formatter heuristics start to get into corners where it takes a lot of expressions interacting to really see what's happening.
